### DISCUSSION TITLE
Persist Data Similar to Preference On Disk

### DISCUSSION THREAD

**Discussion by maheshrajrp** [eclipse-theia/theia#10060] [Q&A] [answered]
Hi,
I'm trying store checkboxes that are selected within a dialog box. So that when the user reopen the dialogue even after closing the browser, he'll be able to get the same checkboxs selected. Is there any recommended way to do the same ? 

The use case is similar to preference, however this doesn't need to be shown in the preference tab and I would like store an array for the preference. 

Thanks for the assist.

**Comment by vince-fugnitto** [comment, marked as answer]
@maheshrajrp you can implement [`StatefulWidget`](https://eclipse-theia.github.io/theia/docs/next/modules/core.statefulwidget-2.html) (which will store the data in local-storage). We use the same idea for many widgets across the framework, for example to store previous input and options selected in the `search-in-workspace-widget`:

- https://github.com/eclipse-theia/theia/blob/master/packages/search-in-workspace/src/browser/search-in-workspace-widget.tsx#L173-L209

For something more persistent you'll probably want to store it on disk.

**Reply by maheshrajrp** [reply]
Hi @vince-fugnitto ,
Thanks for the quick reply. yes, I'm planing to store it on disk. What would be recommendation then ? 

Thanks again.

**Reply by vince-fugnitto** [reply]
@maheshrajrp no real recommendation, you can store it however you like in whichever format you're most comfortable in (ex: json). You'll probably want to save it with a `user-storage` scheme so it is ultimately saved with other `theia` configuration files that gets mapped to `~/.theia/`.

**Reply by maheshrajrp** [reply]
> @maheshrajrp no real recommendation, you can store it however you like in whichever format you're most comfortable in (ex: json). You'll probably want to save it with a `user-storage` scheme so it is ultimately saved with other `theia` configuration files that gets mapped to `~/.theia/`.

I'm having trouble understanding what you mean by "saving it with user-storage scheme". Can you please elaborate on that ? If possible can you provide sample files so that I might be able to refer. 

Thanks

**Reply by vince-fugnitto** [reply]
You can take a look at `@theia/userstorage` and references to it (`usestorage` scheme):

- https://github.com/eclipse-theia/theia/tree/master/packages/userstorage

> The @theia/userstorage extension provides an API for accessing user storage files, i.e preferences, custom keymaps and other user-specific files without having to know how it's implemented. User storage files can then be accessed with userstorage uri scheme like so userstorage://settings.json and the user storage service implementation will then fetch the appropriate file.

**Reply by maheshrajrp** [reply]
Thanks. I will look into this.

**Reply by maheshrajrp** [reply]
@vince-fugnitto I'm creating a class similar to that of keymaps-service.ts to save such data using userstorage. I hope that correct ? Or do you have a better recommendation.

**Reply by vince-fugnitto** [reply]
@maheshrajrp should be fine as a reference, we store `keymaps.json` under the user's home `.theia` for custom commands to keybindings. You'll of course need a way to store your data, retrieve it and so on. The implementation is up to you :)

**Comment by maheshrajrp** [comment]
@msujew  And thanks for your quick reply too. I am currently trying to implement similar to what you mentioned.

**Reply by msujew** [reply]
Well, I deleted it already, because I liked @vince-fugnitto's answer way more 😄 Anyway, here it is for archival purposes:

> Hi, the easiest way is probably to simply store the value of the checkbox in the localStorage of the browser.
However, I'm wondering what you're trying to accomplish here. Preference-like values probably belong inside of the actual preferences. Perhaps it would be best to invoke the preference service when using that dialog and storing the value inside of the service.

**Reply by maheshrajrp** [reply]
Thanks !!!!
